File: Data Structures & Algorithms/squares-of-a-sorted-array/submission-1.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int left = 0, right = nums.size() - 1;
        vector<int> res(nums.size(), 0);
        for (int p = nums.size() - 1; p >= 0; --p) {
            if (nums[left]*nums[left] >= nums[right]*nums[right]) {
                res[p] = nums[left]*nums[left];
                ++left;
            } else {
                res[p] = nums[right]*nums[right];
                --right;
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/sort-an-array/submission-1.py
# import random
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def partition(left, right):
            tmp = nums[left+(right-left)//2]
            nums[left+(right-left)//2] = nums[left]
            # tmp = nums[left]
            while left < right:
                while left < right and nums[right] > tmp:
                    right -= 1
                nums[left] = nums[right]
                while left < right and nums[left] <= tmp:
                    left += 1
                nums[right] = nums[left]
                
            nums[left] = tmp
            return left

        def quiksort(left, right):
            # left close, right close
            if left > right:
                return

            mid = partition(left, right)
            quiksort(left, mid-1)
            quiksort(mid+1, right)

        quiksort(0, len(nums)-1)
        return nums
            
            